storage/internal/elastic/store.go

```diff
@@ -0,0 +1,370 @@
+	// Extract indice name and deploymentID by parsing the key
+	storeType, deploymentID := extractStoreTypeAndDeploymentID(k)
+	indexName := getIndexName(s.cfg, storeType)
+	log.Debugf("storeType is: %s, indexName is: %s, deploymentID is: %s", storeType, indexName, deploymentID)
+
+	// The lastIndex is query by using ES aggregation query ~= MAX(iid) HAVING deploymentId AND clusterId
+	query := buildLastModifiedIndexQuery(s.clusterID, deploymentID)
+	log.Debugf("buildLastModifiedIndexQuery is : %s", query)
+
+	res, err := s.esClient.Search(
+		s.esClient.Search.WithContext(context.Background()),
+		s.esClient.Search.WithIndex(indexName),
+		s.esClient.Search.WithSize(0),
+		s.esClient.Search.WithBody(strings.NewReader(query)),
+	)
+	if err != nil {
+		log.Printf("ERROR: %s", err)
+	}
+	defer res.Body.Close()
+
+	if res.IsError() {
+		var e map[string]interface{}
+		if err := json.NewDecoder(res.Body).Decode(&e); err != nil {
+			log.Printf("error parsing the response body: %s", err)
+		} else {
+			// Print the response status and error information.
+			log.Printf("[%s] %s: %s",
+				res.Status(),
+				e["error"].(map[string]interface{})["type"],
+				e["error"].(map[string]interface{})["reason"],
+			)
+		}
+	}
+
+	var r lastIndexResponse
+	if err := json.NewDecoder(res.Body).Decode(&r); err != nil {
+		log.Printf("Error parsing the response body: %s", err)
+		return 0, err
+	}
+
+	hits := r.hits.total
+	var lastIndex uint64 = 0
+	if hits > 0 {
+		lastIndex, err = parseInt64StringToUint64(r.aggregations.logsOrEvents.lastIndex.value)
+		if err != nil {
+			return lastIndex, err
+		}
+	}
+
+	// Print the response status, number of results, and request duration.
+	log.Debugf(
+		"[%s] %d hits; lastIndex: %d",
+		res.Status(),
+		hits,
+		lastIndex,
+	)
+
+	return lastIndex, nil
+
+}
+
+// List simulates long polling request by :
+// - periodically querying ES for documents (Aggregation to get the max iid and 0 size result).
+// - if a some result is found, wait some time (es_refresh_wait_timeout) in order to:
+//   	- let ES index recently added documents AND to let
+// 		- let Yorc eventually Set a document that has a less iid than the older known document in ES (concurrence issues)
+// - if no result if found after the the given 'timeout', return empty slice
+func (s *elasticStore) List(ctx context.Context, k string, waitIndex uint64, timeout time.Duration) ([]store.KeyValueOut, uint64, error) {
+	log.Debugf("List called k: %s, waitIndex: %d, timeout: %v", k, waitIndex, timeout)
+	if err := utils.CheckKey(k); err != nil {
+		return nil, 0, err
+	}
+
+	// Extract indice name by parsing the key
+	storeType, deploymentID := extractStoreTypeAndDeploymentID(k)
+	indexName := getIndexName(s.cfg, storeType)
+	log.Debugf("storeType is: %s, indexName is: %s, deploymentID is: %s", storeType, indexName, deploymentID)
+
+	query := getListQuery(s.clusterID, deploymentID, waitIndex, 0)
+
+	now := time.Now()
+	end := now.Add(timeout - s.cfg.esRefreshWaitTimeout)
```

Comment:
What if timeout is 0 ?

Reply:
Will substract `s.cfg.esRefreshWaitTimeout` to `now`.
At a first glance it looks good to me.

Reply:
Ok

Reply:
Yes in this case, we'll pass the condition now.After(end) so no timeout at all